operator/pkg/common/extensions/secret_reconciliator.go

```diff
@@ -38,58 +37,57 @@ func (r *SecretReconciliator) Client() ctrlClient.Client {
+	}
+	return nil
 }
 
-// ReconcileSecret reconciles a secret with the given name, by making sure its existence matches "shouldExist" value.
+// ReconcileSecret reconciles a secret with the given name, by making sure it exists.
 // If the validateSecretDataFunc returns an error, then this function calls generateSecretDataFunc to get new secret data and updates the secret to "fix" it.
 // If fixExisting is set to false, an already existing secret will never be overwritten. This is useful only when
 // changing an invalid secret can bring more harm than good.
 // In this case this function will return an error if "validate" rejects the secret.
 // Also note that (regardless of the value of "fixExisting") this function will never touch a secret which is not owned by the object passed to the constructor.
-func (r *SecretReconciliator) ReconcileSecret(ctx context.Context, name string, shouldExist bool, validate validateSecretDataFunc, generate generateSecretDataFunc, fixExisting bool) error {
+func (r *SecretReconciliator) ReconcileSecret(ctx context.Context, name string, validate validateSecretDataFunc, generate generateSecretDataFunc, fixExisting bool) error {
 	secret := &coreV1.Secret{}
-	key := ctrlClient.ObjectKey{Namespace: r.Namespace(), Name: name}
+	key := ctrlClient.ObjectKey{Namespace: r.obj.GetNamespace(), Name: name}
 	if err := r.Client().Get(ctx, key, secret); err != nil {
 		if !apiErrors.IsNotFound(err) {
 			return errors.Wrapf(err, "checking existence of %s secret", name)
 		}
 		secret = nil
 	}
-	if !shouldExist {
-		if secret == nil || !metav1.IsControlledBy(secret, r.obj) {
-			return nil
-		}
-
-		if err := utils.DeleteExact(ctx, r.Client(), secret); err != nil && !apiErrors.IsNotFound(err) {
-			return errors.Wrapf(err, "deleting %s secret", name)
-		}
-		return nil
-	}
 
 	if secret != nil {
 		isManaged := metav1.IsControlledBy(secret, r.obj)
-		var validateErr error
-		if validate != nil {
-			validateErr = validate(secret.Data, isManaged)
-		}
+		validateErr := validate(secret.Data, isManaged)
 
 		if validateErr == nil {
 			return nil // validation of existing secret successful - no reconciliation needed
 		}
-		// If the secret is unmanaged, we cannot fix it, so we should fail. The same applies if there is no
-		// generate function specified, or if the caller told us not to attempt to fix it.
-		if !isManaged || generate == nil || !fixExisting {
+		// If the secret is unmanaged, we cannot fix it, so we should fail. The same applies if
+		// the caller told us not to attempt to fix it.
+		if !isManaged || !fixExisting {
 			return errors.Wrapf(validateErr, "existing %s secret is invalid, please delete the secret to allow fixing the issue", name)
 		}
 	}
 
-	if generate == nil {
-		return pkgUtils.ShouldErr(errors.Errorf("secret %s should exist, but no generation logic has been specified", name))
-	}
-
 	// Try to generate the secret, in order to fix it.
 	data, err := generate()
```

Comment:
Without the guard for nil will this crash now on a nil generator?

Reply:
Discussed offline, clarified that crashing is fine in this case, since `generate` is the only way to provide a value.